core/src/main/java/org/apache/gravitino/storage/relational/service/TableColumnMetaService.java

```diff
@@ -0,0 +1,131 @@
+
+  private TableColumnMetaService() {}
+
+  public static TableColumnMetaService getInstance() {
+    return INSTANCE;
+  }
+
+  public List<ColumnPO> getColumnsByTableIdAndVersion(Long tableId, Long version) {
+    List<ColumnPO> columnPOs =
+        SessionUtils.getWithoutCommit(
+            TableColumnMapper.class,
+            mapper -> mapper.listColumnPOsByTableIdAndVersion(tableId, version));
+
+    // Filter out the deleted columns
+    return columnPOs.stream()
+        .filter(c -> !Objects.equals(c.getColumnOpType(), ColumnPO.ColumnOpType.DELETE.value()))
+        .collect(Collectors.toList());
+  }
+
+  public void insertColumnPOs(TablePO tablePO, List<ColumnEntity> columnEntities) {
+    List<ColumnPO> columnPOs =
+        POConverters.initializeColumnPOs(tablePO, columnEntities, ColumnPO.ColumnOpType.CREATE);
+    SessionUtils.doWithoutCommit(
+        TableColumnMapper.class, mapper -> mapper.insertColumnPOs(columnPOs));
+  }
+
+  public boolean deleteColumnsByTableId(Long tableId) {
+    Integer result =
+        SessionUtils.doWithCommitAndFetchResult(
+            TableColumnMapper.class, mapper -> mapper.softDeleteColumnsByTableId(tableId));
+    return result > 0;
+  }
+
+  public int deleteColumnsByLegacyTimeline(Long legacyTimeline, int limit) {
+    return SessionUtils.doWithoutCommitAndFetchResult(
+        TableColumnMapper.class,
+        mapper -> mapper.deleteColumnPOsByLegacyTimeline(legacyTimeline, limit));
+  }
+
+  public boolean isColumnUpdated(TableEntity oldTable, TableEntity newTable) {
+    Map<Long, ColumnEntity> oldColumns =
+        oldTable.columns() == null
+            ? Collections.emptyMap()
+            : oldTable.columns().stream()
+                .collect(Collectors.toMap(ColumnEntity::id, Function.identity()));
+
+    Map<Long, ColumnEntity> newColumns =
+        newTable.columns() == null
+            ? Collections.emptyMap()
+            : newTable.columns().stream()
+                .collect(Collectors.toMap(ColumnEntity::id, Function.identity()));
+
+    return oldColumns.size() != newColumns.size() || !oldColumns.equals(newColumns);
+  }
+
+  public void updateColumnPOsFromTableDiff(
+      TableEntity oldTable, TableEntity newTable, TablePO newTablePO) {
+    Map<Long, ColumnEntity> oldColumns =
+        oldTable.columns() == null
+            ? Collections.emptyMap()
+            : oldTable.columns().stream()
+                .collect(Collectors.toMap(ColumnEntity::id, Function.identity()));
+    Map<Long, ColumnEntity> newColumns =
+        newTable.columns() == null
+            ? Collections.emptyMap()
+            : newTable.columns().stream()
+                .collect(Collectors.toMap(ColumnEntity::id, Function.identity()));
+
+    List<ColumnPO> columnPOsToInsert = Lists.newArrayList();
+    for (ColumnEntity newColumn : newColumns.values()) {
+      ColumnEntity oldColumn = oldColumns.get(newColumn.id());
+      if (oldColumn == null || !oldColumn.equals(newColumn)) {
```

Comment:
if `oldColumn == null`, it means the newColumn is a newly created one, so the op type should not be update

Reply:
It depends on how we define the `UPDATE` or `CREATE`, here in my implementation, I only set the type to `CREATE` when table is created, `UPDATE` for table update, no matter it is adding new column or updating the existing column.

Actually, no matter it is `UPDATE` or `CREATE`, what we only care if `DELETE`, which is used to filter out deleted columns.